### PR TITLE
feat/union-filter

### LINKED ISSUE #145: Implement UnionFilter
Given two nonequal types `a` and `b`, create their union `a | b`.
Make sure to handle the resulting VARTYPE_MODULE columns properly (hint: if `a` and `b` are from different modules, then we have to be careful :smile: )

### PR DESCRIPTION
**Pull request by bengsparks**
Closes #145

### CODE REVIEW

**Comment by bengsparks** [comment]
Updated to WIP: the import generator to deal with the change to the module format

File: constants.py [hunk 1]
```diff
@@ -64,6 +59,23 @@ class TraceData:
...
+    FILENAME = "Filename"
+    CLASS_MODULE = "ClassModule"
+    CLASS = "Class"
+    FUNCNAME = "FunctionName"
+    LINENO = "LineNo"
+    CATEGORY = "Category"
+    VARNAME = "VarName"
+    VARTYPE_MODULE = "TypeModule"
+    VARTYPE = "Type"
+
```
Comment (RecurvedBow) [inline constants.py:72]: Consider moving all columns into one class and all schema into one class.

For example: constants.Schemas.TypeHintData, constants.Schemas.TraceData, constants.Schemas.Columns.FILENAME, ...

File: tests/resource/configs/example.toml [hunk 2]
```diff
@@ -39,4 +39,8 @@ kind = "keep_only_first"
 [[unifier]]
 name = "min_threshold"
 kind = "drop_min_threshold"
-min_threshold = 0.3
\ No newline at end of file
+min_threshold = 0.3
+
+[[unifier]]
+name = "union-name"
```
Comment (RecurvedBow) [inline tests/resource/configs/example.toml:45]: _ instead of - for consistency.

File: tests/tracing/test_ptconfig.py [hunk 3]
```diff
@@ -62,6 +62,10 @@ def test_full():
     assert config.unifier[7].kind == "drop_min_threshold"
     assert abs(config.unifier[7].min_threshold - 0.3) < 1e-8
 
+    assert isinstance(config.unifier[8], ptconfig.Unify)
+    assert config.unifier[8].name == "union-name"
```
Comment (RecurvedBow) [inline tests/tracing/test_ptconfig.py:66]: Also here

File: tests/typegen/evaluation/test_file_type_hints_collector.py [hunk 4]
```diff
@@ -6,96 +6,146 @@
...
 
 
 def test_file_type_hints_collector_returns_correct_data_for_filename():
     filename1 = str(relative_sample_folder_path / "file_with_type_hints.py")
     filename2 = "file_with_type_hints.py"
 
     test_object = FileTypeHintsCollector()
-    method_to_tests_to_test = [lambda: test_object.collect_data_from_file(root, filename1),
-                               lambda: test_object.collect_data_from_file(sample_folder_path, filename2)]
+    method_to_tests_to_test = [
```
Comment (RecurvedBow) [inline tests/typegen/evaluation/test_file_type_hints_collector.py:32]: I think I accidentally named this wrongly :) Please change all occurences in this file to methods_to_test.

File: typegen/__init__.py [hunk 5]
```diff
@@ -112,7 +112,9 @@ def main(**params):
     collector = TraceDataFileCollector()
     collector.collect_data(traced_df_folder, include_also_files_in_subdirectories=True)
 
-    print(collector.trace_data)
+    td_df = collector.trace_data.astype(constants.TraceData.SCHEMA)
+    print(td_df)
+
```
Comment (RecurvedBow) [inline typegen/__init__.py:118]: Drop one line.

File: typegen/unification/subtyping.py [hunk 6]
```diff
@@ -64,26 +72,28 @@ def _update_group(self, entire: pd.DataFrame, group):
...
                 return group
 
-        group[constants.TraceData.VARTYPE_MODULE] = basetype_module
-        group[constants.TraceData.VARTYPE] = basetype
-        return group
+        updateable = group.copy()
+
+        updateable[TraceData.VARTYPE_MODULE] = basetype_module
+        updateable[TraceData.VARTYPE] = basetype
+        return updateable
```
Comment (RecurvedBow) [inline typegen/unification/subtyping.py:87]: rename to updated_group.

File: typegen/unification/union.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    def _update_group(self, group):
+        if group.shape[0] == 1:
+            module = group[TraceData.VARTYPE_MODULE].values[0]
+            vartype = group[TraceData.VARTYPE].values[0]
+            logger.debug(
+                f"No union to build from module {module}, type {vartype}: Only one value in this group"
+            )
+            return group
+
+        new_module = ",".join(group[TraceData.VARTYPE_MODULE].fillna(""))
```
Comment (RecurvedBow) [inline typegen/unification/union.py:51]: Why do you use different separators?

**Review by RecurvedBow — CHANGES_REQUESTED** [review]

File: tests/typegen/evaluation/test_file_type_hints_collector.py (hunk 4 above)
Comment (bengsparks) [reply tests/typegen/evaluation/test_file_type_hints_collector.py:32]: Err, I don't really want to change your test code. Please fix this after I have fixed your other comments

File: tests/resource/configs/example.toml (hunk 2 above)
Comment (bengsparks) [reply tests/resource/configs/example.toml:45]: Done

File: tests/tracing/test_ptconfig.py (hunk 3 above)
Comment (bengsparks) [reply tests/tracing/test_ptconfig.py:66]: Done

File: typegen/unification/union.py (hunk 7 above)
Comment (bengsparks) [reply typegen/unification/union.py:51]: Because I prepared the VARTYPEs with the union syntax

File: typegen/unification/subtyping.py (hunk 6 above)
Comment (bengsparks) [reply typegen/unification/subtyping.py:87]: Done

File: typegen/__init__.py (hunk 5 above)
Comment (bengsparks) [reply typegen/__init__.py:118]: Done

**Review by RecurvedBow — APPROVED** [review]

**Pull request merged**